Design sidebar

sonicyeti opened the pull request:
Initial PR covering the integration of the latest sidebar design into both SB and the application.

sonicyeti committed:
updated lock file

sonicyeti committed:
Rolling sidebar from SB into application

sonicyeti committed:
Type fix on the method label output

develohpanda commented on packages/insomnia-components/components/sidebar.js:
```diff
@@ -140,204 +213,637 @@ const StyledPanel: React.ComponentType<{}> = styled(motion.div)`
+  // Section Expansion & Filtering
+  // Info
+  const [infoSec, setInfoSec] = useState(false);
+  const toggleInfoSec = () => setInfoSec(!infoSec);
```
Quick note: these callbacks will always create a new instance every time, causing the child components to update unnecessarily due to a prop change, even though most of the `toggle` callbacks will remain unchanged.

I suggest the `useCallback` [hook ](https://reactjs.org/docs/hooks-reference.html#usecallback) to set these, which returns a memoized callback if the dependencies (in this case, `infoSec` and `setInfoSec`) don't change. It's a little more verbose, but you can write your own hook to wrap that logic, eg:

```js
const useToggle = (state, set) => React.useCallback(() => set(!state), [set, state]);
```

Resulting in 
```suggestion
  const toggleInfoSec = useToggle(infoSec, setInfoSec);
```

develohpanda commented on packages/insomnia-components/components/sidebar.js:
```diff
@@ -140,204 +213,637 @@ const StyledPanel: React.ComponentType<{}> = styled(motion.div)`
+      )}
     </StyledSidebar>
   );
 }
```
Is suggest to break this component down in several ways:
1. By each logical section. Components like `<SidebarServers servers={props.spec.servers} />`. Each component can then contain it's respective state and dispatch values, isolated.
2. By the styled components, and the children you expect inside each styled component. A good example would be `StyledFilter` (see comment below).
3. As a nice to have, basic stories for each `<Sidebar* />` component in addition to the all-in-one.

I think you've done an amazing job with how it visually looks. With this breakdown, it will be much easier to unit test, debug and review, and will remove a lot of the code duplication.

Thoughts?

sonicyeti replied:
This sounds great, will start breaking these down now. Originally the plan was one item type to rule them all but we switched it mid-way where we have things like methods grouped/tucked under paths & results. The above will be much easier to work with.

develohpanda commented on packages/insomnia-components/components/sidebar.js:
```diff
@@ -140,204 +213,637 @@ const StyledPanel: React.ComponentType<{}> = styled(motion.div)`
+                onChange={e => setSchemaFilter(e.target.value.toLowerCase())}
+                placeholder="Filter..."
+              />
+            </StyledFilter>
```
Sections like this are duplicated a lot, with the only difference being the value (`schemasFilter` vs `headerFilter`) and the `onChange` handle. Perfect candidate for a component.

The `StyledFilter` is a filter component with only styling. A `SidebarFilter` can be made, which contains a `StyledFilter`, as well as the actual filtering logic and children.

This way, when composing the sidebar itself, adding a filter will be a one-liner:
```
<SidebarFilter filter={headerFilter} onChange={setHeaderFilter} />
<SidebarFilter filter={schemasFilter} onChange={setSchemaFilter} />
```

develohpanda commented on packages/insomnia-components/components/sidebar.js:
```diff
@@ -140,204 +213,637 @@ const StyledPanel: React.ComponentType<{}> = styled(motion.div)`
+              />
+            </StyledFilter>
+            {servers.map((server, index) => (
+              <React.Fragment key={index}>
```
Index as key is a [cause for concern]( https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318), we should definitely look to fix this. Some other instances of it in this components too.

ES Lint should have picked it up, [this rule](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-array-index-key.md) may not be enabled.

develohpanda commented on packages/insomnia-components/components/sidebar.js:
```diff
@@ -140,204 +213,637 @@ const StyledPanel: React.ComponentType<{}> = styled(motion.div)`
-                <SvgIcon icon={IconEnum.plus} />
-              </Tooltip>
-            </motion.span>
+            <Dropdown renderButton={({ open }) => <SvgIcon icon={IconEnum.ellipsesCircle} />}>
```
I believe this syntax will result in this drop-down component and it's children updating on every render. The inline function will mean a new instance is sent every time. Definitely a candidate to declare the a button component outside the component, and consume it here. Also, looks like the `open` prop isn't being used so I have removed it. Eg:
```
const DropdownElipsis = () => <SvgIcon icon={IconEnum.ellipsesCircle} />;
```

```suggestion
            <Dropdown renderButton={DropdownElipsis}>
```

develohpanda commented on packages/insomnia-components/components/sidebar.js:
```diff
@@ -140,204 +213,637 @@ const StyledPanel: React.ComponentType<{}> = styled(motion.div)`
+  const toggleServersSec = () => setServersSec(!serversSec);
+  const [serversFilter, setServersFilter] = useState(false);
+  const toggleServersFilter = () => setServersFilter(!serversFilter);
+  const [serverFilter, setServerFilter] = useState('');
```
I wonder if the state declarations might benefit from `useReducer` instead of `useState`. Easier to test, and less verbose.

develohpanda reviewed (changes requested):
Had a play around in the storybook and it looks amazing. I have added some notes regarding the implementation side of it. The rest are open for discussion, however we should address index as keys in _this_ PR.

gschier commented on packages/insomnia-components/components/sidebar.js:
```diff
@@ -4,16 +4,23 @@ import { useState } from 'react';
 
 type Props = {|
   className?: string,
+  jsonData: Object,
```
We're going to need some sort of event handlers here. I suggest an `onClick` to start that returns an array containing the path to the object that was clicked. 

```js
onClick: (specPath: Array<string | number>) => void,
```

For example, clicking on an operation (request) might call the `onClick` handler with `[ 'paths', '/foo', 'get' ]`

gschier reviewed (changes requested):
@develohpanda did a great job reviewing this! I just had one other comment about adding a click handler so we can hook up the click-to-scroll later.

sonicyeti committed:
Update packages/insomnia-components/components/sidebar.js

Co-authored-by: Opender Singh <opender94@gmail.com>

sonicyeti committed:
PR Feedback - Keys, useCallback Hook

sonicyeti committed:
stubbing component breakout

sonicyeti committed:
PR Feedback

- Breaking out most UI elements into components
- Filter component in place
- Various CSS tweaks

sonicyeti committed:
Removing unused file

sonicyeti committed:
Updating toggle

develohpanda committed:
Resuable sidebar section (#2359)

* Introduce resuable sidebar section
* Stop eslint complaining
* add react-use
* Lint fixes
* Make the storybook story less wide

sonicyeti committed:
Component development

Abstracting out line level components i.e.
Section > Header > Filter > Panel > Items

Various code fixes

sonicyeti committed:
Cleaning up section visibility state

sonicyeti committed:
Reverting visibiilty state updater

sonicyeti committed:
Adding paths on click for spec scrolling

sonicyeti committed:
refreshing package-lock to resolve conflict

develohpanda commented on packages/insomnia-app/app/ui/components/spec-editor/spec-editor-sidebar.js:
```diff
@@ -2,9 +2,11 @@
 import type { ApiSpec } from '../../../models/api-spec';
 import { trackEvent } from '../../../common/analytics';
-import SpecEditorSidebarItem from './spec-editor-sidebar-item';
+const jsyaml = require('js-yaml');
```
Interesting, is this not able to use ES6 imports?

develohpanda commented on packages/insomnia-app/app/ui/components/spec-editor/spec-editor-sidebar.js:
```diff
@@ -36,9 +38,14 @@ class SpecEditorSidebar extends React.PureComponent<Props, State> {
+  componentDidMount() {
+    // const { apiSpec } = this.props;
+  }
+
```
Can be removed

develohpanda commented on packages/insomnia-app/app/ui/components/spec-editor/spec-editor-sidebar.js:
```diff
@@ -118,13 +125,22 @@ class SpecEditorSidebar extends React.PureComponent<Props, State> {
+      specJSON = YAML.parse(apiSpec.contents);
+    } else {
+      specJSON = jsyaml.safeLoad(apiSpec.contents);
+    }
```
Curious about this - could you please elaborate on why we need jsyaml given we only call it if the content type is _not_ yaml?

develohpanda commented on packages/insomnia-app/app/ui/css/components/spec-editor-sidebar.less:
```diff
@@ -1,4 +1,5 @@
 .spec-editor-sidebar {
+  /*
```
Should the commented lines be removed altogether?

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-filter.js:
```diff
@@ -0,0 +1,48 @@
+    initial={{ height: filter ? '100%' : '0px' }}
+    animate={{ height: filter ? '100%' : '0px' }}>
+    <input type="text" placeholder="Filter..." onChange={onChange} />
+    {children}
```
Do any consumers of `SidebarFilter` currently set `children`? If not, we should remove this prop.

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-header.js:
```diff
@@ -0,0 +1,66 @@
+  toggleSection: Function,
+  toggleFilter?: Function,
+  section: boolean,
+  children?: React.Node,
```
- Given we are in a header component, the `headerTitle` prop can simply be `title`
- Is it possible to type the `toggleSection` and `toggleFilter` props more explicitly than `Function`? How about something like `toggleSection: () => void`
- A prop simply named `section` doesn't convey what it represents - I suggest something more explicit like bodyVisible/sectionVisible/sectionExpanded etc.

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar.js:
```diff
@@ -0,0 +1,224 @@
+      set(!state);
+    },
+    [set, state],
+  );
```
Can be removed in favor of `useToggle` in `react-use`. All of the state declarations in this file can simplify, see usage in `SidebarSection`.

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar.js:
```diff
@@ -0,0 +1,224 @@
+  const paths = Object.entries(props.jsonData.paths);
+
+  return (
+    <StyledSidebar className="theme--sidebar">
```
Feels a little strange to use a styled component and also a theme class - what is the reason for using both?

sonicyeti replied:
This is to pickup our inline styling back on preview-head.html, where we essentially dump all CSS variables from core theming. Not an ideal solution and temporary until we centralize all shared styling variables.

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar.js:
```diff
@@ -0,0 +1,224 @@
+// @flow
```
Worth renaming this file to `index.js` given it's in the `sidebar` folder

develohpanda commented on package.json:
```diff
@@ -68,5 +68,7 @@
   },
-  "dependencies": {}
+  "dependencies": {
+    "react-use": "^15.3.2"
```
Doesn't need to be in the root `package.json`

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-header.js:
```diff
@@ -0,0 +1,66 @@
+    <h6 onClick={toggleSection}>{headerTitle}</h6>
+    <div>
+      {children}
+      {!children && (
```
```suggestion
      {children || (
```

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar.js:
```diff
@@ -0,0 +1,224 @@
+import SvgIcon, { IconEnum } from '../svg-icon';
+
+type Props = {|
+  className?: string,
```
`className` prop is not used

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar.js:
```diff
@@ -0,0 +1,224 @@
+  }
+  h5 {
+    font-size: var(--font-size-sm);
+  }
```
None of these elements exist directly in this component; they exist in child components - I suggest these styles be moved closer to where they are used.

sonicyeti replied:
I was able to move them all except svg, which spans header, section, panels and items. Header & section are flat from a DOM perspective due to the header being the target to hide/show panels while persisting visibly.

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-servers.js:
```diff
@@ -0,0 +1,42 @@
+  servers: Array<any>,
+};
+
+let itemPath = [];
+const handleClick = items => {
+  itemPath.push('server');
+  itemPath.push.apply(itemPath, items);
+  console.log(itemPath);
+  itemPath = [];
+};
```
This on click handler, and the element it is linked to, is duplicated through each sidebar section component. Did you want to make a stab at extracting it to a component and reusing? 😄

sonicyeti replied:
Yeah, this was the start of the separate path reveal to the app's code editor wrapper class. This will not be present next round!

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-servers.js:
```diff
@@ -0,0 +1,42 @@
+    <div>
+      {this.props.servers.map(server => (
+        <React.Fragment key={server.url}>
+          {server.url.includes(filter) && (
```
It would be nice to have a placeholder in the empty state (eg. "no results"), which would require a small refactor here. The empty state could also be handled by the `SidebarSection`, so lets discuss some ideas on how to go about doing that.

Currently, this is what the empty state looks like. 

<img width="404" alt="image" src="https://user-images.githubusercontent.com/4312346/87275847-2603ab00-c533-11ea-92b0-f1167276a694.png">

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-servers.js:
```diff
@@ -0,0 +1,42 @@
+        <React.Fragment key={server.url}>
+          {server.url.includes(filter) && (
+            <SidebarItem>
+              <div></div>
```
Rogue empty div? Same with the other components

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-security.js:
```diff
@@ -0,0 +1,42 @@
+    <div>
+      {Object.keys(this.props.security).map(scheme => (
+        <React.Fragment key={scheme}>
+          {scheme.toLowerCase().includes(filter) && (
```
Filtering should be case-insensitive (same with the other components)

<img width="401" alt="image" src="https://user-images.githubusercontent.com/4312346/87276208-fb662200-c533-11ea-868d-c41143f6dabe.png">

<img width="399" alt="image" src="https://user-images.githubusercontent.com/4312346/87276230-04ef8a00-c534-11ea-80d0-c3b959b21cd0.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/4312346/87276257-13d63c80-c534-11ea-8b58-7a27d8607dd8.png">
(no results)

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-info.js:
```diff
@@ -0,0 +1,60 @@
+};
+// Implemented as a class component because of a caveat with render props
+// https://reactjs.org/docs/render-props.html#be-careful-when-using-render-props-with-reactpurecomponent
+export default class SidebarInfo extends React.Component<Props> {
```
Can be a function component, the code comment does not apply to this component

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-item.js:
```diff
@@ -0,0 +1,43 @@
+
+const SidebarItem = ({ children }: Props) => <StyledItem>{children}</StyledItem>;
+
+export default SidebarItem;
```
This component is just a proxy into `StyledItem`, but that means places `SidebarItem` is consumed, there is an empty div to consume the first grid column.

In order to consume `SidebarItem`, one shouldn't need to consider how the `SidebarItem` is styled, so with that in mind, lets discus some ways to abstract that away. 😄 

Same goes for the `SidebarBlockItem`.

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-info.js:
```diff
@@ -0,0 +1,60 @@
+            </span>
+          </SidebarBlockItem>
+        )}
+        {this.props.info.description && (
+          <SidebarBlockItem>
+            <strong>Description:</strong>
+            <span onClick={() => handleClick([this.props.info.description])}>
+              {this.props.info.description}
+            </span>
+          </SidebarBlockItem>
+        )}
```
Destructuring props, and extracting a reusable component in this file would make this really concise 😄

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-panel.js:
```diff
@@ -0,0 +1,24 @@
+import { motion } from 'framer-motion';
+
+type Props = {
+  parent: boolean,
```
What does `parent` represent? I suggest to name the prop according to what it controls in _this_ component. What this prop is linked to, is the responsibility of the consumer. 😄

sonicyeti replied:
Parent wrapper visibility, yeah this could be clearer for sure.

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-requests.js:
```diff
@@ -0,0 +1,75 @@
+export default class SidebarRequests extends React.Component<Props> {
+  renderBody = (filter: string) => (
+    <div>
+      {Object.keys(this.props.requests).map((requestName, value) => (
```
This component can again be broken down, for readability if not for reuse. 👍

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-requests.js:
```diff
@@ -0,0 +1,75 @@
+                    ).map(requestExample => (
+                      <div
+                        onClick={() => handleClick([requestExample])}
+                        className="method-post"
```
Should this className be hard-coded? Feels like it should be linked to the request method 🤔

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-paths.js:
```diff
@@ -0,0 +1,57 @@
+    <div>
+      {this.props.paths.map(path => (
+        <React.Fragment key={path[0]}>
+          {path[0].includes(filter) && (
```
`path[0]` and `path[1]` are difficult to understand, given they are typed as `any`.

I suggest setting the required array items to named variables and consuming those.

`const [foo, bar] = path;`

develohpanda reviewed (changes requested):
Nice one, it's looking at lot more concise, and it's nice to see some separation of concerns where the logic for rendering the requests is scoped, and the logic for rendering servers is scoped, etc.

A little more to go, many of the review comments added to one component apply to other components as well. There is a fair bit of functionality we can still abstract away and extract into reusable components, for readability and making the consumption of each component easier.

sonicyeti committed:
PR feedback for sidebar components

sonicyeti committed:
Update packages/insomnia-components/components/sidebar/sidebar-header.js

Co-authored-by: Opender Singh <opender94@gmail.com>

sonicyeti committed:
Including lock

sonicyeti committed:
Merge branch 'design-sidebar' of https://github.com/Kong/insomnia into design-sidebar

sonicyeti committed:
Merge fix

sonicyeti committed:
Component structuring, prop clean-up, css updates

sonicyeti committed:
Cleaning up components, pr feedback, fixing styling

sonicyeti committed:
array index reference clean-up/clarity

sonicyeti committed:
Leveraging useToggle rather than manual useState toggling

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-info.js:
```diff
@@ -1,49 +1,37 @@
-          <strong>Title:</strong>
-          <span onClick={() => handleClick([title])}>{title}</span>
-        </SidebarBlockItem>
+    <SidebarPanel childrenVisible={props.childrenVisible}>
```
Since all of the children in `SidebarPanel` (title, description, version, license.name) are conditional, if _none_ of them are present an empty panel would render.

What happens/should happen in that case? Should we be rendering `null` (ie. nothing) or an empty `SidebarPanel`?

sonicyeti replied:
Great point, I'm going to back some overarching checks for top level sections into sidebar itself to we don't have empty panels.

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-panel.js:
```diff
@@ -4,8 +4,8 @@ import styled from 'styled-components';
 type Props = {
-  parent: boolean,
   children: React.Node,
+  childrenVisible: boolean,
```
🔥

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar.stories.js:
```diff
@@ -22,7 +22,7 @@ const apiSpec = {
   servers: [
     {
-      url: 'https://dev.konghq.com/v1',
+      url: 'https://dev.kongonghq.com/v1',
```
😆

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-requests.js:
```diff
@@ -18,56 +18,69 @@ const handleClick = items => {
+    return (
+      <div>
+        {filteredValues.map(requestName => (
+          <React.Fragment key={requestName}>
```
I suggest destructing the request data up here.

```js
const {description, content} = this.props.requests[requestName];
...
<Tooltip message={description} ... />
...
{Object.keys(content).map(format => (
    ...
    {Object.keys(content[format].examples).map(requestExample => (...))}
    ...
  )
}
```

sonicyeti replied:
So much cleaner

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-responses.js:
```diff
@@ -20,27 +20,36 @@ const handleClick = items => {
+
+    return (
+      <div>
+        {filteredValues.map(response => (
+          <React.Fragment key={response}>
+            {response.toLowerCase().includes(filter.toLocaleLowerCase()) && (
+              <SidebarItem>
```
This can be simplified now

```suggestion
        {filteredValues.map(response => (
          <SidebarItem key={response}>
```

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-schemas.js:
```diff
@@ -19,23 +19,32 @@ const handleClick = items => {
+
+    return (
+      <div>
+        {filteredValues.map(schema => (
+          <React.Fragment key={schema}>
+            {schema.toLowerCase().includes(filter.toLocaleLowerCase()) && (
+              <SidebarItem>
```
```suggestion
        {filteredValues.map(schema => (
          <SidebarItem key={schema}>
```

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-paths.js:
```diff
@@ -17,39 +17,53 @@ const handleClick = items => {
+      <div>
+        {filteredValues.map(([route, method]) => (
+          <React.Fragment key={route}>
+            {route.includes(filter.toLocaleLowerCase()) && (
```
This check should be in the filter above on line 29 and not here, I think?

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-section.js:
```diff
@@ -35,9 +40,11 @@ const SidebarSection = ({ title, renderBody }: SectionProps) => {
         <SidebarFilter filter={filterVisible} onChange={handleFilterChange} />
-        {renderBody(filterValue)}
+        {renderBody(filterValue) || (
+          <StyledNoResults>No results found for "{filterValue}"...</StyledNoResults>
```
👍

develohpanda reviewed (approved):
Couple of things worth addressing, either in this PR or a separate one, otherwise good to go. 👍

sonicyeti committed:
Cleaning up logs

sonicyeti committed:
Hide info section if missing from spec.

sonicyeti committed:
Update packages/insomnia-components/components/sidebar/sidebar-responses.js

Co-authored-by: Opender Singh <opender94@gmail.com>

sonicyeti committed:
Update packages/insomnia-components/components/sidebar/sidebar-schemas.js

Co-authored-by: Opender Singh <opender94@gmail.com>

sonicyeti committed:
Pull fragment

sonicyeti committed:
Destructuring and filter cleanup

sonicyeti committed:
Clearing filter value and DOM on section collapse

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-section.js:
```diff
@@ -32,6 +33,11 @@ const SidebarSection = ({ title, renderBody }: SectionProps) => {
     setFilterValue(e.target.value);
   }, []);
 
+  useLayoutEffect(() => {
+    toggleFilterVisible(false);
+    setFilterValue('');
+  }, [bodyVisible, toggleFilterVisible]);
```
Nice! I had to research about the difference between `useEffect` and `useLayoutEffect`, [TIL](https://daveceddia.com/useeffect-vs-uselayouteffect/)! 🤗

develohpanda commented on packages/insomnia-components/components/sidebar/sidebar-filter.js:
```diff
@@ -35,12 +36,22 @@ const StyledFilter: React.ComponentType<{}> = styled(motion.div)`
-  </StyledFilter>
-);
+const SidebarFilter = ({ filter, onChange }: Props) => {
+  const filterField = React.createRef();
```
There is a `useRef` hook to use here in place of `React.createRef()`. Although I don't know the difference between them, would favor `useRef` to stay with hooks.

It would also be nice to achieve this without using a ref altogether, but I'm not totally sure what the solution might be right now (maintaining the value as state), so not going to suggest removing the ref yet. 😄

develohpanda reviewed (approved):
Nice one 👍

sonicyeti committed:
Merge branch 'develop' into design-sidebar

sonicyeti committed:
Merge latest

sonicyeti committed:
Merge conflict cleanup
